File: scripts/harness-skill-link.mjs
#!/usr/bin/env node
// ECL-HARNESS-CONNECTOR
// Attach or detach this worktree from its project-level shared Harness Skill without requiring Python.

import fs from "node:fs";
import path from "node:path";
import { execFileSync } from "node:child_process";
import process from "node:process";

const skillName = "agent-harness-orchestrator-a6ad344cbe4e-harness";
const projectId = "agent-harness-orchestrator-a6ad344cbe4e";

function git(cwd, ...args) {
  return execFileSync("git", ["-C", cwd, ...args], { encoding: "utf8" }).trim();
}

function readItem(itemPath) {
  try {
    return fs.lstatSync(itemPath);
  } catch (error) {
    if (error?.code === "ENOENT") return null;
    throw error;
  }
}

function normalizePath(itemPath) {
  const normalized = path.resolve(itemPath).replace(/[\\/]+$/, "");
  return process.platform === "win32" ? normalized.toLowerCase() : normalized;
}

function sameTarget(link, target) {
  try {
    return normalizePath(fs.realpathSync.native(link)) === normalizePath(fs.realpathSync.native(target));
  } catch {
    return false;
  }
}

function rejectLinkedAncestors(root, itemPath) {
  const relative = path.relative(path.resolve(root), path.dirname(path.resolve(itemPath)));
  if (relative === ".." || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative)) {
    throw new Error(`Skill path escapes this worktree: ${itemPath}`);
  }
  let current = path.resolve(root);
  for (const part of relative.split(path.sep).filter(Boolean)) {
    current = path.join(current, part);
    const item = readItem(current);
    if (item?.isSymbolicLink()) {
      throw new Error(`Skill path must not traverse a link or junction: ${current}`);
    }
    if (item && !item.isDirectory()) {
      throw new Error(`Skill path parent must be a directory: ${current}`);
    }
  }
}

function addSkillLink(root, link, target) {
  if (normalizePath(link) === normalizePath(target)) return "physical";
  rejectLinkedAncestors(root, link);
  const item = readItem(link);
  if (item) {
    if (item.isSymbolicLink() && sameTarget(link, target)) return "existing";
    throw new Error(`Skill path collision: ${link}`);
  }
  fs.mkdirSync(path.dirname(link), { recursive: true });
  const type = process.platform === "win32" ? "junction" : "dir";
  const value = process.platform === "win32" ? target : path.relative(path.dirname(link), target);
  fs.symlinkSync(value, link, type);
  return "attached";
}

function removeSkillLink(link) {
  fs.unlinkSync(link);
}

function resolveProject() {
  const current = process.cwd();
  const root = path.resolve(git(current, "rev-parse", "--show-toplevel"));
  const commonRaw = git(root, "rev-parse", "--git-common-dir");
  const common = path.isAbsolute(commonRaw) ? path.resolve(commonRaw) : path.resolve(root, commonRaw);
  const worktreeLines = git(root, "worktree", "list", "--porcelain").split(/\r?\n/);
  const firstWorktree = worktreeLines.find((line) => line.startsWith("worktree "));
  const primary = path.basename(common) === ".git"
    ? path.dirname(common)
    : path.resolve(firstWorktree?.slice("worktree ".length).trim() ?? "");
  if (!primary) throw new Error("could not resolve the primary worktree");
  return { root, primary };
}

function validateCanonical(primary) {
  const canonical = path.join(primary, ".agents", "skills", skillName);
  rejectLinkedAncestors(primary, canonical);
  const item = readItem(canonical);
  if (!item || (!item.isDirectory() && !item.isSymbolicLink())) {
    throw new Error(`canonical project Harness is missing: ${canonical}`);
  }
  if (item.isSymbolicLink()) {
    throw new Error(`canonical project Harness must be physical: ${canonical}`);
  }
  if (!fs.statSync(path.join(canonical, "SKILL.md"), { throwIfNoEntry: false })?.isFile()) {
    throw new Error(`canonical project Harness is missing: ${canonical}`);
  }
  const manifest = JSON.parse(fs.readFileSync(path.join(canonical, "state", "manifest.json"), "utf8"));
  if (manifest.project_id !== projectId || manifest.skill_name !== skillName) {
    throw new Error("canonical project Harness manifest does not match this Git project");
  }
  return canonical;
}

function attach(root, canonical, links) {
  const result = {};
  const created = [];
  try {
    for (const [name, link] of Object.entries(links)) {
      const status = addSkillLink(root, link, canonical);
      result[name] = { path: link, status };
      if (status === "attached") created.push(link);
    }
  } catch (error) {
    for (const link of created.reverse()) {
      try { removeSkillLink(link); } catch { /* Preserve the original connector error. */ }
    }
    throw error;
  }
  return result;
}

function detach(root, primary, canonical, links) {
  if (normalizePath(root) === normalizePath(primary)) {
    throw new Error("the primary worktree hosts the physical project Harness and cannot be detached");
  }
  const result = {};
  for (const [name, link] of Object.entries(links)) {
    rejectLinkedAncestors(root, link);
    const item = readItem(link);
    if (!item) {
      result[name] = { path: link, status: "missing" };
    } else if (!item.isSymbolicLink()) {
      throw new Error(`refusing to detach an unmanaged physical Skill path: ${link}`);
    } else if (!sameTarget(link, canonical)) {
      throw new Error(`refusing to detach a Skill link with the wrong target: ${link}`);
    } else {
      result[name] = { path: link, status: "detached" };
    }
  }

  const removed = [];
  try {
    for (const [name, link] of Object.entries(links)) {
      if (result[name].status !== "detached") continue;
      removeSkillLink(link);
      removed.push(link);
    }
  } catch (error) {
    const rollbackErrors = [];
    for (const link of removed.reverse()) {
      try { addSkillLink(root, link, canonical); } catch (rollbackError) {
        rollbackErrors.push(`${link}: ${rollbackError.message}`);
      }
    }
    const detail = rollbackErrors.length > 0 ? `; rollback failed for ${rollbackErrors.join(", ")}` : "";
    throw new Error(`could not detach all shared Harness links: ${error.message}${detail}`);
  }
  return result;
}

const args = process.argv.slice(2);
if (args.some((arg) => arg !== "--detach") || args.filter((arg) => arg === "--detach").length > 1) {
  throw new Error("usage: harness-skill-link.mjs [--detach]");
}
const shouldDetach = args.includes("--detach");
const { root, primary } = resolveProject();
const canonical = validateCanonical(primary);
const links = {
  codex: path.join(root, ".agents", "skills", skillName),
  claude: path.join(root, ".claude", "skills", skillName),
};
const result = shouldDetach
  ? detach(root, primary, canonical, links)
  : attach(root, canonical, links);
process.stdout.write(`${JSON.stringify({
  ok: true,
  action: shouldDetach ? "detached" : "attached",
  skill: canonical,
  links: result,
}, null, 2)}\n`);

File: scripts/harness-skill-link.py
#!/usr/bin/env python3
# ECL-HARNESS-CONNECTOR
"""Attach or detach this worktree's shared project Harness Skill links."""

from __future__ import annotations

import argparse
import json
import os
import subprocess
from pathlib import Path


SKILL_NAME = "agent-harness-orchestrator-a6ad344cbe4e-harness"
PROJECT_ID = "agent-harness-orchestrator-a6ad344cbe4e"


def git(*args: str, cwd: Path) -> str:
    result = subprocess.run(
        ["git", "-C", str(cwd), *args],
        text=True,
        encoding="utf-8",
        errors="replace",
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        check=False,
    )
    if result.returncode != 0:
        raise RuntimeError(result.stderr.strip() or "git command failed")
    return result.stdout.strip()


def is_link_like(path: Path) -> bool:
    is_junction = getattr(os.path, "isjunction", None)
    return path.is_symlink() or bool(is_junction and is_junction(path))


def path_exists(path: Path) -> bool:
    return os.path.lexists(path)


def normalized(path: Path) -> str:
    return os.path.normcase(os.path.abspath(path))


def same_target(path: Path, target: Path) -> bool:
    try:
        return normalized(path.resolve(strict=True)) == normalized(target.resolve(strict=True))
    except OSError:
        return False


def reject_linked_ancestors(root: Path, path: Path) -> None:
    root = Path(os.path.abspath(root))
    parent = Path(os.path.abspath(path.parent))
    try:
        relative = parent.relative_to(root)
    except ValueError as exc:
        raise RuntimeError(f"Skill path escapes this worktree: {path}") from exc
    current = root
    for part in relative.parts:
        current = current / part
        if path_exists(current) and is_link_like(current):
            raise RuntimeError(f"Skill path must not traverse a link or junction: {current}")
        if path_exists(current) and not current.is_dir():
            raise RuntimeError(f"Skill path parent must be a directory: {current}")


def link_directory(root: Path, path: Path, target: Path) -> str:
    if normalized(path) == normalized(target):
        return "physical"
    reject_linked_ancestors(root, path)
    if path_exists(path):
        if is_link_like(path) and same_target(path, target):
            return "existing"
        raise RuntimeError(f"Skill path collision: {path}")
    path.parent.mkdir(parents=True, exist_ok=True)
    if os.name == "nt":
        result = subprocess.run(
            ["cmd", "/c", "mklink", "/J", str(path), str(target)],
            text=True,
            encoding="utf-8",
            errors="replace",
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            check=False,
        )
        if result.returncode != 0:
            raise RuntimeError(result.stderr.strip() or f"could not create junction {path}")
    else:
        path.symlink_to(os.path.relpath(target, path.parent), target_is_directory=True)
    return "attached"


def remove_link(path: Path) -> None:
    if os.name == "nt" and not path.is_symlink():
        os.rmdir(path)
    else:
        path.unlink()


def resolve_project() -> tuple[Path, Path]:
    current = Path.cwd().resolve()
    root = Path(git("rev-parse", "--show-toplevel", cwd=current)).resolve()
    common_raw = Path(git("rev-parse", "--git-common-dir", cwd=root))
    common = common_raw.resolve() if common_raw.is_absolute() else (root / common_raw).resolve()
    worktrees = git("worktree", "list", "--porcelain", cwd=root).splitlines()
    primary_line = next((line for line in worktrees if line.startswith("worktree ")), None)
    if common.name == ".git":
        primary = common.parent.resolve()
    elif primary_line:
        primary = Path(primary_line.removeprefix("worktree ").strip()).resolve()
    else:
        raise RuntimeError("could not resolve the primary worktree")
    return root, primary


def validate_canonical(primary: Path) -> Path:
    canonical = primary / ".agents" / "skills" / SKILL_NAME
    reject_linked_ancestors(primary, canonical)
    if not path_exists(canonical) or not canonical.is_dir():
        raise RuntimeError(f"canonical project Harness is missing: {canonical}")
    if is_link_like(canonical):
        raise RuntimeError(f"canonical project Harness must be physical: {canonical}")
    if not (canonical / "SKILL.md").is_file():
        raise RuntimeError(f"canonical project Harness is missing: {canonical}")
    manifest = json.loads((canonical / "state" / "manifest.json").read_text(encoding="utf-8"))
    if manifest.get("project_id") != PROJECT_ID or manifest.get("skill_name") != SKILL_NAME:
        raise RuntimeError("canonical project Harness manifest does not match this Git project")
    return canonical


def attach(root: Path, canonical: Path, links: dict[str, Path]) -> dict[str, dict[str, str]]:
    result: dict[str, dict[str, str]] = {}
    created: list[Path] = []
    try:
        for name, path in links.items():
            status = link_directory(root, path, canonical)
            result[name] = {"path": str(path), "status": status}
            if status == "attached":
                created.append(path)
    except Exception:
        for path in reversed(created):
            try:
                remove_link(path)
            except OSError:
                pass
        raise
    return result


def detach(
    root: Path,
    primary: Path,
    canonical: Path,
    links: dict[str, Path],
) -> dict[str, dict[str, str]]:
    if normalized(root) == normalized(primary):
        raise RuntimeError("the primary worktree hosts the physical project Harness and cannot be detached")
    result: dict[str, dict[str, str]] = {}
    for name, path in links.items():
        reject_linked_ancestors(root, path)
        if not path_exists(path):
            result[name] = {"path": str(path), "status": "missing"}
        elif not is_link_like(path):
            raise RuntimeError(f"refusing to detach an unmanaged physical Skill path: {path}")
        elif not same_target(path, canonical):
            raise RuntimeError(f"refusing to detach a Skill link with the wrong target: {path}")
        else:
            result[name] = {"path": str(path), "status": "detached"}

    removed: list[Path] = []
    try:
        for name, path in links.items():
            if result[name]["status"] != "detached":
                continue
            remove_link(path)
            removed.append(path)
    except Exception as error:
        rollback_errors = []
        for path in reversed(removed):
            try:
                link_directory(root, path, canonical)
            except Exception as rollback_error:
                rollback_errors.append(f"{path}: {rollback_error}")
        detail = f"; rollback failed for {', '.join(rollback_errors)}" if rollback_errors else ""
        raise RuntimeError(f"could not detach all shared Harness links: {error}{detail}") from error
    return result


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--detach", action="store_true")
    args = parser.parse_args()
    root, primary = resolve_project()
    canonical = validate_canonical(primary)
    links = {
        "codex": root / ".agents" / "skills" / SKILL_NAME,
        "claude": root / ".claude" / "skills" / SKILL_NAME,
    }
    result = detach(root, primary, canonical, links) if args.detach else attach(root, canonical, links)
    print(json.dumps({
        "ok": True,
        "action": "detached" if args.detach else "attached",
        "skill": str(canonical),
        "links": result,
    }, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
